### ISSUE TITLE
Move ediff to evil-collection

### ISSUE THREAD

**Issue by Emiller88** [emacs-evil/evil-collection#129]
Just wondering why evil-ediff hasn't been added here. 

If no one sees any issue I'll open a PR.

**Comment by jojojames** [comment]
Related to https://github.com/emacs-evil/evil-collection/issues/28

We can revisit this topic if there's enough interest. As of right now, I'm ok with leaving the packages separate but there's a couple options to bundle them together (requiring evil-ediff when installing evil-collection, etc).

Not sure how others feel about this though.

**Comment by Emiller88** [comment]
That was what I was looking for was that there was already a conversation. I was just interested why there was an evil-collection-magit, and an evil-magit.

I probably don't understand the history of the emacs-evil org.

**Comment by jojojames** [comment]
Yeah, I think it'd be pretty cool if there was just one package users could import (evil-collection) and all their evil-integration woes would be handled. In that manner, probably requiring (adding those packages as dependencies might be interesting). I haven't really thought too much of the downsides though.

I'd be curious what @Ambrevar and @noctuid thinks.

**Comment by noctuid** [comment]
Because `evil-collection` intends to provide consistent keybindings across modes, I think it makes sense to have all modes supported, at least eventually. In some cases, it might make sense to ask other package authors to merge their packages into evil-collection. Alternatively, evil-collection could just depend on them like you suggest and potentially reconcile any inconsistent keybindings. For packages that only make keybindings and don't do any other form of integration, this would be less necessary.

One thing to consider is whether other packages make sense to integrate here, at least in their entirety (specifically I'm thinking about evil-org). Ediff and mu4e support definitely makes sense. I don't think it would make sense to integrate evil-lispy or lispyville at all. Finally, org agenda keybindings definitely make sense, but org mode keybindings don't fit with the general theme (keybindings in non-editing modes where a lot of letter keys can be rebound), so only some of evil-org's functionality is relevant to evil-collection.

**Comment by Ambrevar** [comment]
I don't see much of a problem with evil-org: after all, many other modes supported by Evil Collection do not bear the same "consistency" because of their inherent structure that is too different (Helm, diff, etc.).

I don't know about evil-lispy, but lispyville seems to be a tad unfinished as of now.  We've been discussing binding themes and much of the bindings seem to be extremely opinionated.  Much more than the rest of Evil Collection.

At last, evil-mu4e is also unfinished in my opinion but much more doable.  I've been working on polishing it and I think I'll send a patch soon.

**Comment by noctuid** [comment]
Well evil-collection hardly binds anything in diff-mode when the buffer isn't read only. I'd consider some of the keybindings in the org packages fairly opinionated. I don't necessarily think it's a bad idea add support for "normal" editable modes like org mode or programming modes; there is room for consistent keybindings between markup modes, between programming modes, etc. I think it would make sense to put most of the functionality on a customizable prefix key at that point. In a mode like org mode where there are less keys available for use and a lot of potentially useful commands, I think user setups vary a lot more. Right now evil-collection doesn't do much about normal modes. If you wanted to support some, I think it would be best to either not have that as the default or to have a variable containing editing modes, so that users could easily remove them from `evil-collection-mode-list` if they only wanted to use evil-collection for non-editing modes/applications.

Even once lispyville is more finished, I think lispyville keybindings are too opinionated for it to make sense to add them here.

**Comment by justbur** [comment]
I'm happy to merge evil-ediff into this repo, and you're welcome to make changes. I'm not using it much these days anyway. 

I would also merge evil-magit if you want, although evil-magit is more than a "consistency" layer because so many decisions have to be made within it. 

I would also like to suggest that some basic documentation in the form of a README for each mode would be helpful when you're browsing this package on github (each mode in a folder with an associated README). This could list the bindings changed in a table (plus perhaps the important bindings left in place), explain choices made if necessary. I would hate to lose the evil-magit documentation if it were merged for example. 

Going a little further it would be nice to have a function like `evil-collection-describe-mode` that popped up the same documentation within emacs.

**Comment by Ambrevar** [comment]
I tend to dislike README and prefer contextual documentation, and I think you are totally right, Evil Collection is in urgent need of a proper "mode description function".
It might be a little difficult to get right, though.

I suggest we go all in with the contextual description feature so that the READMEs of Ediff and Magit (among others) can be moved there.

**Comment by Ambrevar** [comment]
I've sent a [PR](https://github.com/JorisE/evil-mu4e/pull/14) to evil-mu4e.
It's not completely finished yet but nearly there.  Feedback welcome! :)

**Comment by jojojames** [comment]
> One thing to consider is whether other packages make sense to integrate here, at least in their entirety (specifically I'm thinking about evil-org).
> Ediff and mu4e support definitely makes sense.
> I don't think it would make sense to integrate evil-lispy or lispyville at all.
> Finally, org agenda keybindings definitely make sense, but org mode keybindings don't fit with the general theme (keybindings in non-editing modes where a lot of letter keys can be rebound), so only some of evil-org's functionality is relevant to evil-collection.

> If you wanted to support some, I think it would be best to either not have that as the default or to have a variable containing editing modes, so that users could easily remove them from evil-collection-mode-list if they only wanted to use evil-collection for non-editing modes/applications.
> Even once lispyville is more finished, I think lispyville keybindings are too opinionated for it to make sense to add them here.

Generally agree with this though I happen to use both evil-org and lispyville.

> I'm happy to merge evil-ediff into this repo, and you're welcome to make changes. I'm not using it much these days anyway.

Sounds good. 

> I would also merge evil-magit if you want, although evil-magit is more than a "consistency" layer because so many decisions have to be made within it.

I think that would be awesome!

> I would also like to suggest that some basic documentation in the form of a README for each mode would be helpful when you're browsing this package on github (each mode in a folder with an associated README). This could list the bindings changed in a table (plus perhaps the important bindings left in place), explain choices made if necessary. I would hate to lose the evil-magit documentation if it were merged for example.
> Going a little further it would be nice to have a function like evil-collection-describe-mode that popped up the same documentation within emacs.

This sounds reasonable to me.

> I tend to dislike README and prefer contextual documentation, and I think you are totally right, Evil Collection is in urgent need of a proper "mode description function".
> It might be a little difficult to get right, though.

Maybe we can have both? Sometimes I like reading the README through a web browser.

On that note, do members of the evil org already have write access to this repo?
At the very least @justbur should have write access once those packages are merged.

Taking a quick look, I'm surprised @Ambrevar and @noctuid aren't in the evil org already.

**Comment by Somelauw** [comment]
> One thing to consider is whether other packages make sense to integrate here, at least in their entirety (specifically I'm thinking about evil-org).

I think evil-org is quite big compared to the other packages and has some optional features that don't necessarily follow conventions of evil-collecion, but have been proven useful to some.

> On that note, do members of the evil org already have write access to this repo?

I would be okay with adding evil-org to the emacs-evil organisation somehow.

**Comment by jojojames** [comment]
> On that note, do members of the evil org already have write access to this repo?

>> I would be okay with adding evil-org to the emacs-evil organisation somehow.

That would be good I think. @Ambrevar Personally I'd invite @noctuid @justbur @Somelauw as collaborators with commit access to evil-collection but I think now that evil-collection has moved under emacs-evil, that permission/setting is out of my control.

**Comment by Ambrevar** [comment]
Done.

**Comment by jojojames** [comment]
By the way, thanks for bringing this up @Emiller88!

**Comment by syl20bnr** [comment]
My 2 cents about moving `evil-magit` into `evil-collection`. 

TL;DR It would be a terrible decision for Spacemacs to do so but I believe it would also be a bad decision for `evil-collection`.

`evil-magit` is a complex package with configuration variables and a lot of stuff done behind the scene to make it integrated flawlessly with `magit`. If this project should be moved somewhere it would be better to be moved under the `magit` flagship IMO but I guess this won't happen.

From an `evil-collection` point of view now. If this package is move then you'll have to maintain good documentation for the users which may not be as trivial. `which-key` does the job for key-bindings discovery but it won't do it for configuration variables etc...

As I already stated, any kitchen-sink package is unfortunately not well suited for Spacemacs as it defeats one of the most important Spacemacs design which is to break down functionalities and isolate them by layers. I find `evil-collection` a wonderful initiative to cover up simple key bindings but it can be harmful in the long run if we throw everything in it, even complex packages that are more than just key rebindings.

Maybe `evil-magit` could be part of `evil-collection` via dependencies instead of assimilation ? So we can let users have more fine-grained control over their configuration.

**Comment by Ambrevar** [comment]
I'm not very knowledgeable about Spacemacs, so could you explain how it makes it more difficult?
We've recently merged evil-ediff and evil-mu4e and beside the namespace adjustment there was basically no change.  Let me know if there is something else.

As for the documentation: some effort was made recently so that we can auto-generate a bindings documentation buffer, which looks much better than which-key (`evil-collection-describe-bindings`), and I suppose it would not be too hard to attach some extra documentation there as well.

The configuration variables should be `defcustom`s and as such they are discoverable and documented.  But this is true to all Emacs packages, not just Evil-magit or evil-collection.

Evil-collection "sub-packages" are mostly independent from each other, so I think it's no trouble if one of them happens to be much more complex.  It won't impact the others.

**Comment by syl20bnr** [comment]
You do what you want :-) But please consider the impact of your moves for the others.

I would highly appreciate if you don't remove another package from MELPA, it will save precious time for me.

For instance `evil-ediff` could have stay in MELPA, there was no point in removing it in the first place. Maybe the first import in`evil-collection` was easy but as time goes by I'm sure `evil-ediff` will converge more and more toward `evil-collection` API and conventions, making it a whole new package compared to the original `evil-ediff` repository.

Thank you!
